scripthost/src/main/kotlin/com/commonwealthrobotics/bowlerkernel/scripthost/Session.kt

```diff
@@ -0,0 +1,271 @@
+ * (at your option) any later version.
+ *
+ * bowler-kernel is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with bowler-kernel.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package com.commonwealthrobotics.bowlerkernel.scripthost
+
+import arrow.core.Either
+import com.commonwealthrobotics.bowlerkernel.authservice.Credentials
+import com.commonwealthrobotics.bowlerkernel.authservice.CredentialsProvider
+import com.commonwealthrobotics.bowlerkernel.di.BowlerKernelKoinComponent
+import com.commonwealthrobotics.bowlerkernel.gitfs.GitFS
+import com.commonwealthrobotics.bowlerkernel.gitfs.GitHubFS
+import com.commonwealthrobotics.bowlerkernel.protoutil.sessionServerMessage
+import com.commonwealthrobotics.bowlerkernel.protoutil.withTask
+import com.commonwealthrobotics.bowlerkernel.scripting.Script
+import com.commonwealthrobotics.bowlerkernel.scripting.ScriptLoader
+import com.commonwealthrobotics.bowlerkernel.util.CallbackLatch
+import com.commonwealthrobotics.bowlerkernel.util.toChannel
+import com.commonwealthrobotics.proto.script_host.RunRequest
+import com.commonwealthrobotics.proto.script_host.SessionClientMessage
+import com.commonwealthrobotics.proto.script_host.SessionServerMessage
+import com.google.protobuf.GeneratedMessageV3
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.InternalCoroutinesApi
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.channels.produce
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.consumeAsFlow
+import kotlinx.coroutines.selects.select
+import mu.KotlinLogging
+import org.koin.dsl.module
+import java.util.concurrent.atomic.AtomicLong
+
+/**
+ * Implements the bidirectional session RPC.
+ *
+ * @param coroutineScope The scope used to process messages from the client flow. Cannot be GlobalScope.
+ * @param client The client flow.
+ */
+class Session(
+    private val coroutineScope: CoroutineScope,
+    private val client: Flow<SessionClientMessage>
+) : CredentialsProvider, BowlerKernelKoinComponent {
+
+    // Bind a GitFS using this session so that credentials requests will go through this session
+    private val modules by lazy {
+        listOf(module { single<GitFS> { GitHubFS(this@Session) } })
+    }
+
+    // Session-scoped unique ID counters
+    private val nextRequest = AtomicLong(1)
+    private val nextTaskID = AtomicLong(1)
+
+    // Request-response handling
+    private val responseHandlers: MutableMap<Long, (SessionClientMessage) -> Unit> = mutableMapOf()
+    private val requests = CallbackLatch<GeneratedMessageV3.Builder<*>, SessionClientMessage>()
```

Comment:
Is there a reason why the type is `GeneratedMessageV3.Builder<*>` instead of `SessionServerMessage.Builder`?

Reply:
Nope